hoomd/md/alchemy/pair.py

```diff
@@ -0,0 +1,367 @@
+                                              k)
+            items[k] = self._data[k]
+        if len(items) == 1:
+            return items.popitem()[1]
+        return items
+
+    def __iter__(self):
+        """Iterate over keys."""
+        yield from self._data
+
+    def __contains__(self, key):
+        """Return whether the key is in the mapping."""
+        keys = list(self._indexer(key))
+        if len(keys) == 1:
+            return keys[0] in self._data
+        return [k in self._data for k in keys]
+
+    def __len__(self):
+        """Get the length of the mapping."""
+        return len(self._data)
+
+    def _attach(self, types):
+        self._indexer.valid_types = types
+        for key in self:
+            if not self._indexer.are_valid_types(key):
+                raise RuntimeError(
+                    f"Alchemical DOF ({self._name}) for non-existent type pair "
+                    f"{key} was accessed.")
+
+    def _detach(self):
+        self._indexer.valid_types = None
+
+
+class _AlchemicalPairForce(_HOOMDBaseObject):
+    """Base class for Alchemical pair potentials.
+
+    Expects to use diamond inheritance with a `hoomd.md.pair.Pair` subclass.
+    Automatically creates the `AlchemicalPairDOFStore` objects in `__init__` and
+    implements a type parameter like interface for accessing them.
+
+    Attributes:
+        _alchemical_dofs (`list`[ `str`]): A list of all potential degrees of
+            freedom. This must match that of the ``params`` type parameter.
+        _dof_cls (AlchemicalDOF): The correct DOF class. Automatically set via
+            `_modify_pair_cls_to_alchemical`.
+    """
+    _alchemical_dofs = []
+    _dof_cls = None
+
+    def __init__(self):
+        self._set_alchemical_parameters()
+
+    def _set_alchemical_parameters(self):
+        self._alchemical_params = {}
+        for dof in self._alchemical_dofs:
+            self._alchemical_params[dof] = AlchemicalPairDOFStore(
+                name=dof, pair_instance=self, dof_cls=self._dof_cls)
+
+    def _setattr_hook(self, attr, value):
+        if attr in self._alchemical_dofs:
+            raise RuntimeError(f"{attr} is not settable.")
+        super()._setattr_hook(attr, value)
+
+    def _getattr_hook(self, attr):
+        if attr in self._alchemical_dofs:
+            return self._alchemical_params[attr]
+        return super()._getattr_hook(attr)
+
+
+class AlchemicalDOF(_HOOMDBaseObject):
+    """Alchemical degree of freedom :math:`\\alpha_i` associated with a\
+    specific pair force.
+
+    `AlchemicalDOF` represents an alchemical degree of freedom to be
+    numerically integrated via an `alchemical integration method
+    <hoomd.md.alchemy.methods>`.
+
+    Note:
+        To access an alchemical particle for a particular degree of freedom
+        query the corresponding attribute in the class instance with alchemical
+        degrees of freedom.
```

Comment:
This note is more confusing than helpful as written. Maybe something like this would be better: 

```suggestion
    Note:
        To access an alchemical degree of freedom for a particular type pair,
        query the corresponding attribute in the alchemical pair force instance.
```

Reply:
@joaander feel free to accept suggestion or modify as you see fit and resolve.